site/docs/spec.md

```diff
@@ -375,7 +375,7 @@ The schema of a manifest file is a struct called `manifest_entry` with the follo
 | _optional_ | _optional_ | **`109  value_counts`**           | `map<119: int, 120: long>`   | Map from column id to number of values in the column (including null and NaN values) |
 | _optional_ | _optional_ | **`110  null_value_counts`**      | `map<121: int, 122: long>`   | Map from column id to number of null values in the column |
 | _optional_ | _optional_ | **`137  nan_value_counts`**       | `map<138: int, 139: long>`   | Map from column id to number of NaN values in the column |
-| _optional_ |            | ~~**`111 distinct_counts`**~~     | `map<123: int, 124: long>`   | **Deprecated. Do not write.** |
+| _optional_ | _optional_ | **`111  distinct_counts`**        | `map<123: int, 124: long>`   | Map from column id to number of distinct values in the column; distinct counts must be produced using values in the file, not on merged counts from other metadata |
```

Comment:
We probably don't need this qualification in the spec `not on merged counts from other metadata`.  The spec should be clear enough that it should reflect the distinct count in the file (anything else is out of spec by definition).  You might want to add that if an operation is performed that would compromise the correctness of the value, the field should be dropped.

Reply:
I think it is a bit more clear to state that this can't be a merged count. If someone is compacting data files without reading the data, like merging Parquet row groups without re-encoding then it may seem reasonable to merge the counts somehow. I'm not sure that people would consider merging an operation that doesn't reflect the distinct count in the file, that's why I mentioned it specifically.

Is there anything else that you're including in cases that would not "reflect the distinct count in the file"? I'm trying to use language that allows non-exact values. I think we should be in agreement that sketch-based counts are okay.

Reply:
I agree that it is better to state that it should not be derived by merging counts because it is easy to go towards that route. But it feels to me that the definition is a bit too specific. What I am thinking is to say that:

> distinct counts must be derived using values in the file. This can be done through methods like counting or sketch, but not through merging counts from other metadata.

Reply:
@rdblue I think the line `distinct_counts was removed` in https://github.com/apache/iceberg/blob/072a86a9d83d6c43d7e19cef0bbba80a94714794/site/docs/spec.md#L1073 needs to be removed as well

Reply:
I updated this to the wording that @jackye1995 suggested, but a bit more condensed.

I also removed the line that @nastra found.

Reply:
Should the spec explicitly say "approximate number of distinct values"? It says that implicitly by saying "sketches", but it's not obvious at a glance.